File: 1-100/81-90/86-partitionList/partitionList.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def partition(self, head, x):
        """
        :type head: ListNode
        :type x: int
        :rtype: ListNode
        """
        
        node = head
        leftHead = rightHead = left = right = None

        while node:
            if node.val < x:
                if left:
                    left.next = node
                else:
                    leftHead = node
                left = node
            else:
                if right:
                    right.next = node
                else:
                    rightHead = node
                right = node
            
            node = node.next
        
        if left:
            left.next = rightHead
        if right:
            right.next = None
        return leftHead if leftHead else rightHead

File: 1-100/81-90/86-partitionList/partitionList.cpp

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode *end = NULL, *greaterHead = NULL, *greaterEnd = NULL;

        for (ListNode *lptr = head; lptr; lptr = lptr->next) {
            if (lptr->val < x) {
                if (end)
                    end = end->next = lptr;
                else
                    head = end = lptr;
            }
            else {
                if (greaterEnd)
                    greaterEnd = greaterEnd->next = lptr;
                else
                    greaterHead = greaterEnd = lptr;
            }
        }

        if(greaterEnd)
            greaterEnd->next = NULL;
        if(end)
            end->next = greaterHead;

        return head;
    }
};